First pass on the C Test Generation

thk123 opened the pull request:
Includes support for generating tests for:

- non main methods
- normal parameters being passed in
- struct (and nested struct) parameters and returns
- asserting results on returns

dcattaruzza committed:
get symbol of member

dcattaruzza committed:
Auxiliary function for getting byte offsets of members

dcattaruzza committed:
Auxiliary function to check if the ssa has enough data to build an identifier

dcattaruzza committed:
Bitwise operators for mpinteger

dcattaruzza committed:
Auxiliary function to retrieve tail of trace
Added dead command to trace

dcattaruzza committed:
Changes to interpreter to cover a number of java functions and objects
including:
bitwise operations
variable size arrays
dereferencing
dead objects
dynamical objects
input enumeration (trace backtracking)

Commit:
Basic adding of the gen-c-test-case

Adding to the help, adding a method stuff in cbmc_parse_options when added

Commit:
Creating stub for c_test_case_generator 

Modified the various make files to support the new directory.

Commit:
Created a gitignore to ignore files generated by QTCreator

Commit:
More general gitignore for ignoring qt files

Commit:
Created function for generating test name from the function name

Method body takes the function name, sanitises it and converts it into a test name. Also stubbed a method for generating the body

Commit:
Removing legacy method for calling bmc::run through the test generator

Commit:
Initial helper functions in the factory class for emitting C++ test file

Tracks indentaiton level when creating the file.

Commit:
Adding header guiards to interpreter_class.h

Commit:
Refactored method for getting the entry function into interpreter

The get_entry_function was required by both the C test gen stuff (to name the test) and in the interpreter to generate the counter example inputs.

Commit:
Added doxygen comment to last function

Commit:
Create a set of parameters corresponding to the inputs for a function

Commit:
Adding function for checking whether to skip over a variable

Checks the name for the CPROVER prefix

Commit:
Added function for recoding variables inside the symbol_table

When calling evaluate_address, for symbols, if not found we add a new entry in to the memory model. To get this compiling it involved making a bunch of stuff mutable so need to look at this at some point.

Commit:
Extended the factory to using input_varst to fill out the parameter list

Currently using a simple approach to deducing the variable values (rather than using the interpretert::load_counter_example_inputs)

Commit:
Pass in the file name so can add an appropriate include

Currently getting the file name from the first goal which isn't the best way perhaps

Commit:
Fixing emitted print statement

Commit:
Made the C Test generator find the file name of the funciton

Rather than using arbitarily the first goal, we deduce the file name from where the entry function is declared (which is what we need to include to call it). Also moved logic into the C Test Generator since specific to C.

Commit:
Merge branch 'cristina-david/interpreter' into 'c-test-gen'

Commit:
Fixing make file to get to compile 

Was missing JSON dependency in CBMC for some reason

Commit:
Simple dumping of the generated file to output for now

Commit:
Swapping to using the interpreter to get the input values

Currently not working correctly as includes some junk inputs that we need to filter out. Also produces quite a lot of errors in the log

Commit:
Added function for printing a trace in an informative-ish way

Commit:
Correcting file from previous commit that wasn't saved

Commit:
Using my simple refactor of identifying CPROVER variables

Commit:
Using expr2c to more generally create variable inputs

Also started overriding functions in expr2c in a expr2cleanc which will create compilable C. This particular one means when assigning to a struct, we don't assign the padding variables which don't really exist.

Commit:
Handling in the expr2cleanc not outputting the body of struct

The body of the struct can be turned off (or optionally just the padding variables)

Commit:
Adding simple regression for struct function

Also made test logs be ignored but git. This test currently fails (correctly) as the integer member of the struct is not being set to the correct value

Commit:
Added regression for when a function has a local variable

Again currently correctly failing as trying to pass the local variable in as a parameter

Commit:
Added basic regression for very simple function

Commit:
Made my make clean rerun make after removing the log

Commit:
Fixed the structs in the clean C generator

Was incorrectly using the value of the padding variable (0) to initalise the next one as we weren't stepping the iterators in step.

Commit:
Made booleans be lower case when generating clean C

This requires including stdbool.h but still more correct than capitalised TRUE/FALSE.

Commit:
Fixed reorder warning

dcattaruzza committed:
Interpreter class header guards
Avoid crash from infinite array size

Commit:
Merge remote-tracking branch 'refs/remotes/cristina-david/interpreter'

Conflicts:
	src/goto-programs/interpreter_class.h

dcattaruzza committed:
_H for header guards

Commit:
Merged branch interpreter into c-test-gen

Commit:
Refactor the get_entry_function into a seperate function 

Allows for getting the expression seperately to the body

Commit:
Refactored print_inputs to allow passing in a different input_vars

Commit:
Filter out inputs that aren't function inputs in the test gen

This is more consistent with how the Java test gen was working and there were some issues with the prune_inputs

Commit:
Adding simple regression test for returns

Commit:
Corrected regression test

Commit:
Basic handling of function returns to assert the result

Commit:
Added a warning for when function returns a value but has return void

Also included a simple regression that should produce this warning

Commit:
Adding tests for not-yet-implemented features

Pointers, struct returns and function calls

Commit:
Refactored out the code for the function inputs into a class

Commit:
Refactored out code for generating the return statement

Also pulled out some useful type defs for the input entries

Commit:
Modified the function return builder to support asserts for structs

Currently doesn't support nested structs but does do an assertion per member variable in a struct

Commit:
Made the assertion generation support nested structs

Commit:
Regression test for nested structs as inputs

Commit:
Moved c_test_file into its own file

Commit:
Removing redundant code 

Logic got moved into the function_return_builder

Commit:
Adding doxygen comments for recently added code

Commit:
Adding support for making the test harness not run on main

This can then be compiled with files that include a main.

Commit:
Don't use stdbool to make combatible with some of the regressions

Commit:
Updated regression tests to not use stdbool and support test_main

Commit:
Allowed the expr2cleanc to create array decleration with implict size

Move the array type converter into seperate method with argument specifying whether, if even possible, to include the array size.

Commit:
Added function for converting type with a supplied identifier

This allows for array types to put the identifier between the type of the contents of the array and the square brackets.

Commit:
Added regression tests for arrays and the main function

The array ones fail as don't initalise arrays to anything other than NULL
The main fails as we try to assign &argv ' to params which is at best a type mismatch

Commit:
Adding regression test for typedef'd structs

Bug outlined in #315

Commit:
Merge diffblue/master into c-test-gen-no-prune

peterschrammel commented on src/ansi-c/expr2c.cpp:
```diff
@@ -705,6 +672,85 @@ std::string expr2ct::convert_rec(
+  {
+    dest+=" {";
+
+    for(struct_typet::componentst::const_iterator
```
use ranged for

peterschrammel commented on src/ansi-c/expr2c_class.h:
```diff
@@ -208,6 +228,9 @@ class expr2ct
   std::string convert_concatenation(const exprt &src, unsigned &precedence);
   std::string convert_sizeof(const exprt &src, unsigned &precedence);
+
+  std::string convert_struct(const exprt &src, unsigned &precedence,
```
put each of the parameters on a separate line

thk123 replied:
So I can correctly reapply this, what is the rule here? Is it because one of the parameters is on a new line therefore they all should be?

peterschrammel commented on src/cbmc/Makefile:
```diff
@@ -27,7 +27,9 @@ OBJ += ../ansi-c/ansi-c$(LIBEXT) \
       ../assembler/assembler$(LIBEXT) \
       ../solvers/solvers$(LIBEXT) \
-      ../util/util$(LIBEXT)
+			../test-c-gen/test-c-gen$(LIBEXT) \
```
indentation too large

peterschrammel commented on src/cbmc/bmc_cover.cpp:
```diff
@@ -309,6 +311,42 @@ bool bmc_covert::operator()()
+      std::vector<std::string> goal_names;
+      for(const auto& goalid : test.covered_goals)
+        goal_names.push_back(
+                             as_string(goal_map.at(goalid).description)
```
indent only 2 spaces from current base column

thk123 replied:
As in this line should read like:

    goal_names.push_back(as_string(goal_map.at(goalid).description)+"\n *  "
      +id2string(goal_map.at(goalid).source_location.get_file())
      +":"
      +id2string(goal_map.at(goalid).source_location.get_line()));

And is this rule true for function parameters in headers, e.g. the above `convert_struct` should be formatted:

    std::string convert_struct(const exprt &src,
      unsigned &precedence,
      bool include_padding_members);

peterschrammel commented on src/cbmc/bmc_cover.cpp:
```diff
@@ -309,6 +311,42 @@ bool bmc_covert::operator()()
+      for(const auto& goalid : test.covered_goals)
+        goal_names.push_back(
+                             as_string(goal_map.at(goalid).description)
+                             + "\n *  "
```
no space after +

peterschrammel commented on src/cbmc/bmc_cover.cpp:
```diff
@@ -309,6 +311,42 @@ bool bmc_covert::operator()()
+                             + id2string(goal_map.at(goalid).source_location.get_line()));
+
+      // Compute the test function name
+      /*test.test_function_name=gen.generate_test_func_name(bmc.ns.get_symbol_table(),
```
use #if 0 to deactivate code

peterschrammel commented on src/cbmc/bmc_cover.cpp:
```diff
@@ -309,6 +311,42 @@ bool bmc_covert::operator()()
+                                                   goto_functions,test.goto_trace,
+                                                   ++test_case_no,goal_names);*/
+
+      std::string test_function_name = gen.get_test_function_name(bmc.ns.get_symbol_table(),
```
no spaces around =

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -31,8 +41,41 @@ Function: interpretert::operator()
 void interpretert::operator()()
 {
+  show=true;
+  message->status() << "0- Initialize:" << messaget::eom;
```
interpretert should be derived from messaget

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -31,8 +41,41 @@ Function: interpretert::operator()
 {
+  show=true;
+  message->status() << "0- Initialize:" << messaget::eom;
+  initialise(true);
```
US english for identifiers -> initialize

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -31,8 +41,41 @@ Function: interpretert::operator()
+
+ \*******************************************************************/
+
+void interpretert::initialise(bool init) {
```
opening brace in next line

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -48,12 +91,128 @@ void interpretert::operator()()
+  }
+}
+
+std::set<irep_idt> interpretert::get_parameter_set(
```
missing comment header

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -48,12 +91,128 @@ void interpretert::operator()()
+
+  rev_instruction_itert last_function_call = std::find_if(
+        start_instructions.rbegin(), start_instructions.rend(),
+        [&](const instructiont &instruction ) {
```
opening brace in next line

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -48,12 +91,128 @@ void interpretert::operator()()
+  typedef goto_programt::instructiont instructiont;
+  typedef goto_programt::instructionst::const_reverse_iterator rev_instruction_itert;
+
+  rev_instruction_itert last_function_call = std::find_if(
```
put find_if on next line

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -48,12 +91,128 @@ void interpretert::operator()()
+          user provided function (e.g. the actual function being tested)
+
+\*******************************************************************/
+const code_function_callt &interpretert::get_entry_function_call(const goto_functionst &gf)
```
spell out variable names -> goto_functions

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -48,12 +91,128 @@ void interpretert::operator()()
+const code_function_callt &interpretert::get_entry_function_call(const goto_functionst &gf)
+{
+  typedef goto_functionst::function_mapt function_mapt;
+  const function_mapt &fm = gf.function_map;
```
fm -> function_map

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -48,12 +91,128 @@ void interpretert::operator()()
+\*******************************************************************/
+const code_function_callt &interpretert::get_entry_function_call(const goto_functionst &gf)
+{
+  typedef goto_functionst::function_mapt function_mapt;
```
use a typedef if it is used at least twice; typedefs for a single namespace nesting look exaggerated

thk123 replied:
The reason why I did this was to keep under the 80 character limit for the line. Though for this line this doesn't seem to be a problem. Would you transform this line:

    const goto_functionst::function_mapt::const_iterator start_function = function_map.find(goto_start_id);

To a typedef of const_iterator, or just put a line break after the =?

Another case I was addressing the range based for loop in expr2c.cpp. I got:

    typedef struct_union_typet::componentt componentt;
    for(const componentt &component : struct_type.components())

Without the typedef, it reads like this:

    for(const struct_union_typet::componentt &component : struct_type.components())
                                                                                ^
                                                                   80 characters here

What approach should be taken here to keep under 80 characters?

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -48,12 +91,128 @@ void interpretert::operator()()
+\*******************************************************************/
+const exprt &interpretert::get_entry_function(const goto_functionst &gf)
+{
+  const code_function_callt &func_call = get_entry_function_call(gf);
```
no spaces around =

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -48,12 +91,128 @@ void interpretert::operator()()
+
+\*******************************************************************/
+bool interpretert::get_is_variable_cprover(
+    const irep_idt &variable_id)
```
put on same line if there is space

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -48,12 +91,128 @@ void interpretert::operator()()
+bool interpretert::get_is_variable_cprover(
+    const irep_idt &variable_id)
+{
+  constexpr char const *cprover_variable_prefix = "__CPROVER";
```
this is declared in util/prefix.h

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -48,12 +91,128 @@ void interpretert::operator()()
+
+  const char *cell_identifier = variable_id.c_str();
+
+  if(strncmp(cell_identifier, cprover_variable_prefix, cprover_prefix_len) == 0)
```
use has_prefix() from util/has_prefix.h

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -48,12 +91,128 @@ void interpretert::operator()()
+
+  if(strncmp(cell_identifier, cprover_variable_prefix, cprover_prefix_len) == 0)
+  {
+    return true;
```
if-then-else seems exaggerated here

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -65,25 +224,28 @@ Function: interpretert::show_state
-            << std::endl;
+  if(!show) return;
+  message->status() << messaget::endl;
+  message->status() << total_steps+1 << " ----------------------------------------------------"
```
break line

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -136,27 +411,35 @@ void interpretert::step()
   
   next_PC=PC;
-  next_PC++;  
+  next_PC++; 
```
prefer use of ++next_PC

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -288,15 +618,211 @@ void interpretert::execute_decl()
+      result.copy_to_operands(operand);
+    }
+    return result;
+  } else if(real_type.id()==ID_array) {
```
put else on next line

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -288,15 +618,211 @@ void interpretert::execute_decl()
+
+    result.reserve_operands(components.size());
+    for(struct_typet::componentst::const_iterator it=components.begin();
+        it!=components.end();++it) {
```
brace on next line

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -288,15 +618,211 @@ void interpretert::execute_decl()
+    const struct_typet::componentst &components=struct_type.components();
+
+    result.reserve_operands(components.size());
+    for(struct_typet::componentst::const_iterator it=components.begin();
```
ranged for

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -288,15 +618,211 @@ void interpretert::execute_decl()
+      unsigned size=get_size(it->type());
+      assert(size>=0);
+      const exprt operand=get_value(it->type(), rhs, offset);
+      offset += size;
```
no spaces around +=

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -288,15 +618,211 @@ void interpretert::execute_decl()
+    result.reserve_operands(count);
+    for(unsigned i=0;i<count;i++) {
+      const exprt operand=get_value(type.subtype(), rhs,
+          offset+i * subtype_size);
```
weird spaces

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -288,15 +618,211 @@ void interpretert::execute_decl()
+    }
+
+    return result;
+  } else if(real_type.id()==ID_array) {
```
else and { on next line

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -330,22 +882,28 @@ Function: interpretert::assign
   const std::vector<mp_integer> &rhs)
 {
-  for(unsigned i=0; i<rhs.size(); i++, ++address)
+  for(unsigned i=0;i<rhs.size();i++, ++address)
```
prefer ++i

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -330,22 +882,28 @@ Function: interpretert::assign
+                          << messaget::endl << messaget::eom;
+      }
       cell.value=rhs[i];
+      if(cell.initialised==0) cell.initialised=1;
```
-> initialized

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -430,15 +996,15 @@ void interpretert::execute_function_call()
     return_value_address=0;
```
no indent

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -458,25 +1024,10 @@ void interpretert::execute_function_call()
-      if(size!=0)
-      {
-        frame.local_map[id]=stack_pointer;
+      frame.local_map[id]=integer2unsigned(build_memory_map(id,symbol.type));
```
missing indent

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -564,91 +1143,1266 @@ void interpretert::build_memory_map(const symbolt &symbol)
     memory_map[symbol.name]=address;
     
-    for(unsigned i=0; i<size; i++)
+    for(unsigned i=0;i<size;i++)
```
space after ;

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -564,91 +1143,1266 @@ void interpretert::build_memory_map(const symbolt &symbol)
+  unsigned size=get_size(alloc_type);
+  std::map<const irep_idt,const typet>::iterator it = dynamic_types.find(id);
+
+  if (it!=dynamic_types.end()) {
```
{ on next line

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -564,91 +1143,1266 @@ void interpretert::build_memory_map(const symbolt &symbol)
+
+    unsigned max_size=0;
+
+    for(union_typet::componentst::const_iterator
```
ranged for

peterschrammel commented on src/goto-programs/interpreter.cpp:
```diff
@@ -564,91 +1143,1266 @@ void interpretert::build_memory_map(const symbolt &symbol)
+void interpretert::list_non_bodied() {
+  int funcs=0;
+  function_input_vars.clear();
+  for(goto_functionst::function_mapt::const_iterator f_it =
```
no space before =

peterschrammel commented on src/test-c-gen/c_test_case_generator.cpp:
```diff
@@ -0,0 +1,196 @@
+Outputs: An executable C test harness
+Purpose: To generate a test harness that produces a specific trace
+\*******************************************************************/
+std::string c_test_case_generatort::generate_tests(const optionst &options,
```
put parameters with 2 space indentation on next line if line is too long

peterschrammel commented on src/test-c-gen/c_test_case_generator.cpp:
```diff
@@ -0,0 +1,196 @@
+Purpose: To generate a test harness that produces a specific trace
+\*******************************************************************/
+std::string c_test_case_generatort::generate_tests(const optionst &options,
+                                                   const symbol_tablet &st,
```
spell out names: st -> symbol_table, etc

peterschrammel commented on src/test-c-gen/c_test_case_generator.cpp:
```diff
@@ -0,0 +1,196 @@
+                                                   const size_t test_idx,
+                                                   const std::vector<std::string> &goals)
+{
+  test_case_generatort generator = generate_c_test_case_from_inputs;
```
no space around =

peterschrammel commented on src/test-c-gen/c_test_case_generator.cpp:
```diff
@@ -0,0 +1,196 @@
+{
+  test_case_generatort generator = generate_c_test_case_from_inputs;
+  return generate_tests_with_generator(options, st, gf, trace, generator,
+                                test_idx, goals);
```
align with first param or put all parameters with 2 spaces indent on next line

peterschrammel commented on src/test-c-gen/c_test_case_generator.cpp:
```diff
@@ -0,0 +1,196 @@
+  const std::string pretty_function_name = as_string(st.lookup(called_function_name).pretty_name);
+  const std::string sanitised_name = sanitise_function_name(pretty_function_name);
+
+  // In the java version we hash the function name and append it, but I don't
```
In java with all the package prefixes, the test name can become very long. In C it should be alright.

peterschrammel commented on src/test-c-gen/c_test_case_generator.cpp:
```diff
@@ -0,0 +1,196 @@
+
+  // In the java version we hash the function name and append it, but I don't
+  // see a reason for this
+  // It also pads the functions with 0's, again, I'm not sure why
```
I think the test index is padded with leading zeros to make it a 3 digit number

thk123 replied:
Is this preferable? I couldn't see the value in it but happy to do if you think it is better.

peterschrammel commented on src/test-c-gen/c_test_case_generator.cpp:
```diff
@@ -0,0 +1,196 @@
+Outputs: A clean version of the name.
+Purpose: Remove ., <, > from names.
+\*******************************************************************/
+std::string c_test_case_generatort::sanitise_function_name(
```
sanitise -> sanitize

peterschrammel commented on src/test-c-gen/c_test_case_generator.cpp:
```diff
@@ -0,0 +1,196 @@
+  // Print out the trace
+
+  status() << "[";
+  std::vector<std::string> trace_types = {"NONE", "ASSIGNMENT", "ASSUME",
```
goto_trace_stept should provide a function for stringifying the enum.

thk123 replied:
I removed this since I think it is redundant given the `--trace` flag

peterschrammel commented on src/test-c-gen/c_test_case_generator.h:
```diff
@@ -0,0 +1,105 @@
+
+\*******************************************************************/
+
+#ifndef C_TEST_CASE_GENERATOR_H
```
add a further prefix CPROVER_

peterschrammel commented on src/test-c-gen/c_test_case_generator.h:
```diff
@@ -0,0 +1,105 @@
+#include <goto-programs/interpreter_class.h>
+
+// This is the full original function type that we're working towards
+/*typedef std::function<
```
use #if 0 to deactivate code

peterschrammel commented on src/test-c-gen/c_test_case_generator.h:
```diff
@@ -0,0 +1,105 @@
+{
+}
+
+  /**
```
no doxy comments in the header file for now

peterschrammel commented on src/test-c-gen/c_test_file.cpp:
```diff
@@ -0,0 +1,195 @@
+#include "c_test_file.h"
```
header missing

peterschrammel commented on src/test-c-gen/c_test_file.cpp:
```diff
@@ -0,0 +1,195 @@
+\*******************************************************************/
+void c_test_filet::end_main_method()
+{
+  add_line_at_current_indentation("exit(0);");
```
why is this needed?

thk123 replied:
To support calling `main` from the test, I was calling the test harnesses main method `test_main` and the compiling with the flags  `-Wl,-etest_main -nostartfiles`. If you do this, then to avoid a segfault you need to call `exit(0)` at the end: http://stackoverflow.com/a/34380445

peterschrammel replied:
ok

peterschrammel commented on src/test-c-gen/c_test_file.cpp:
```diff
@@ -0,0 +1,195 @@
+\*******************************************************************/
+void c_test_filet::add_closing_brace(int level)
+{
+  assert(level == current_indentation - 1);
```
no spaces around operators (except << && ||)

peterschrammel commented on src/test-c-gen/c_test_file.cpp:
```diff
@@ -0,0 +1,195 @@
+\*******************************************************************/
+void c_test_filet::add_line_at_indentation(std::string line, int level)
+{
+  current_file += indentation(level) + line + "\n";
```
no spaces

peterschrammel commented on src/test-c-gen/c_test_source_factory.cpp:
```diff
@@ -0,0 +1,138 @@
+#include <test-c-gen/function_return_builder.h>
+
+
+/*std::string generate_c_test_case_from_inputs(
```
#if 0

peterschrammel commented on src/test-c-gen/c_test_file.h:
```diff
@@ -0,0 +1,41 @@
+#ifndef C_TEST_FILE_H
```
header missing
CPROVER_

peterschrammel commented on src/test-c-gen/c_test_source_factory.cpp:
```diff
@@ -0,0 +1,138 @@
+Purpose: To pull out from the inputs the relevant ones for the function
+         call
+\*******************************************************************/
+  inputst filter_inputs_to_function_parameters(const inputst &all_inputs,
```
put functions inside a class

thk123 replied:
I took the structure from the Java version before I really understood. I've removed the factory file (it just had a bunch of functions relating to generation that were called from the generator so distinction made no sense). 

At the moment my main issue is that we pass a lot of parameters around. Would it be preferable to move the parameters into a constructor and then have a simple call to generate the tests. I've seen this style elsewhere.

peterschrammel commented on src/test-c-gen/c_test_source_factory.h:
```diff
@@ -0,0 +1,59 @@
+ *
+ * @return The C code?
+ */
+std::string generate_c_test_case_from_inputs(
```
refactor this into a class with an operator()

peterschrammel commented on src/test-c-gen/expr2cleanc.cpp:
```diff
@@ -0,0 +1,80 @@
+                                      false, false);
+}
+
+std::string expr2cleanct::convert_array_type(const typet &src,
```
header missing

thk123 replied:
Sorry I don't think I understand - what header is missing?

peterschrammel replied:
the comment header describing the function

peterschrammel commented on src/test-c-gen/expr2cleanc.cpp:
```diff
@@ -0,0 +1,80 @@
+std::string expr2cleanct::convert_constant_bool(bool boolean_value)
+{
+  // This requires #include <stdbool.h>
+  if(boolean_value)
```
maybe use conditional operator

peterschrammel commented on src/test-c-gen/expr2cleanc.h:
```diff
@@ -0,0 +1,38 @@
+
+\*******************************************************************/
+
+#ifndef EXPR2CLEANC_H
```
CPROVER_ prefix

peterschrammel commented on src/test-c-gen/function_parameter_builder.cpp:
```diff
@@ -0,0 +1,65 @@
+#include "function_parameter_builder.h"
```
header missing

peterschrammel commented on src/test-c-gen/function_parameter_builder.cpp:
```diff
@@ -0,0 +1,65 @@
+  std::string type = e2c.convert_with_identifier(entry.second.type(),
+                                                 variable_name);
+  var_assignment_builder << type;
+
```
remove some blank lines

peterschrammel commented on src/test-c-gen/function_parameter_builder.cpp:
```diff
@@ -0,0 +1,65 @@
+}
+
+/*******************************************************************\
+Function:  function_parameter_buildert::get_parameter_decleration
```
decleration -> declaration

peterschrammel commented on src/test-c-gen/function_parameter_builder.h:
```diff
@@ -0,0 +1,29 @@
+
+\*******************************************************************/
+
+#ifndef FUNCTION_PARAMETER_BUILDER_H
```
prefix CPROVER_

peterschrammel commented on src/test-c-gen/function_return_builder.cpp:
```diff
@@ -0,0 +1,203 @@
+/*******************************************************************\
+Function: function_return_buildert::function_return_buildert
+Inputs:
+ all_inputs - All the inputs from the interpreter
```
inputs are those occurring in INPUT steps, outputs are those in OUTPUT steps

thk123 replied:
Sorry, I'm not sure I follow. As in the comment for the constructor should explain all input-y functions within the class?

peterschrammel replied:
No, sorry, this comment seems misplaced. 
There was a more general question in some place how the interpreter knows what inputs and outputs are. Inputs and outputs of the function under test are marked in the goto trace.

peterschrammel commented on src/test-c-gen/function_return_builder.cpp:
```diff
@@ -0,0 +1,203 @@
+  for(const input_entryt &input : all_inputs)
+  {
+    // TODO: this isn't a great way to find returns
+    if(input.first == "return'")
```
the interpreter should find all the variables in OUTPUT steps

peterschrammel commented on src/test-c-gen/function_return_builder.cpp:
```diff
@@ -0,0 +1,203 @@
+         each of them.
+ \*******************************************************************/
+void function_return_buildert::add_assertions_for_struct_expression(
+    const exprt &correct_expression, std::string ret_value_var)
```
this will require some generalisation if we want to generate assertions for side effects too

thk123 replied:
Definitely, should I add an issue to support side effect assertions (at the moment side effects are completely ignored.

peterschrammel commented on src/test-c-gen/function_return_builder.cpp:
```diff
@@ -0,0 +1,203 @@
+  // be called
+  assert(has_return);
+
+  std::ostringstream ret_var_decleartion_builder;
```
-> declaration

peterschrammel commented on src/test-c-gen/function_return_builder.h:
```diff
@@ -0,0 +1,46 @@
+
+\*******************************************************************/
+
+#ifndef FUNCTION_RETURN_BUILDER_H
```
CPROVER_

peterschrammel commented on src/test-c-gen/input_entry.h:
```diff
@@ -0,0 +1,20 @@
+/*******************************************************************
```
not sure why this is needed. Can't the interpreter declare that?

peterschrammel reviewed (commented):
Some issues are recurring - I put a comment only the first time I encountered it.

I also partly commented the interpreter code - I'll have to address these comments to Dario.

Commit:
Tidied up the printing out of the tests to be the same as Java test generation.

Commit:
Adjusted function_buildert to use the type of the LHS rather than the RHS of an expression

Commit:
Removed debug print of trace as functionality already provided by --trace

Commit:
Removed unused function from interpretert

Commit:
Addressing feedback:
- Removing white space
- Removing incorrect doxygen comments
- Full variable names
- Using ranged for in expr2c
- Corrected spelling of declaration & American spellings

Commit:
Moved the function for generating the tests into c_test_case_generator
The division between the two files was unnecessary as using separate
classes to handle specific parts of the generation.

Commit:
More declaration correction

Commit:
Using typedefs from the interpreter

Commit:
Appended CPROVER_ to the header declarations
Corrected includes in source files to use the full module path and angle brackets
Added comment header to header file in c_test_file

Commit:
Corrected indentation of broken lines to be shifted one tab in from the previous lines column

Commit:
Removed single depth typedefs and replaced with line breaks

Commit:
Added function header comment to expr2cleanct::convert_array_type

Commit:
Number the tests with padding so that they will correctly alphabetise
